rules/private/phases/phase_semanticdb.bzl

```diff
@@ -1,5 +1,12 @@
 load("@bazel_skylib//lib:paths.bzl", "paths")
-load("@rules_scala_annex//rules:providers.bzl", _ScalaConfiguration = "ScalaConfiguration")
+load(
+    "@rules_scala_annex//rules:providers.bzl",
+    _ScalaConfiguration = "ScalaConfiguration",
+    _SemanticDbInfo = "SemanticDbInfo",
+)
+
+def _semanticdb_directory_from_file(file):
+    return file.path[:file.path.find("META-INF") - 1]
```

Comment:
I probably should've mentioned this in a comment, but I acknowledge this is pretty janky and could match on the wrong part of the path. Unfortunately, we're pretty limited in what we can do here. From the [documentation](https://bazel.build/rules/lib/builtins/Args#add_all) on `Args#add_all`:

> To avoid unintended retention of large analysis-phase data structures into the execution phase, the `map_each` function must be declared by a top-level `def` statement; it may not be a nested function closure by default.

It does work with path mapping because this function will be called at execution time, not analysis time. In fact, this sort of approach is [recommended](https://github.com/bazelbuild/bazel/discussions/22658) for use with path mapping:
> Generally speaking, all functions that return a path string that may contain a configuration prefix such as bazel-out/darwin-amd64-fastbuild/bin must only be called in map_each callbacks, where they are automatically path mapped by Bazel.

Using `short_path` doesn't work because the compiler expects the output base to be in the path as its working directory is not the output base. I can do this instead, though:
```starlark
def _semanticdb_directory_from_file(file):
    return "{}/{}".format(file.root.path, file.short_path[:file.short_path.find("META-INF") - 1])
```